### PR TITLE
Added the prime_finder function and it's unittest #30

### PR DESCRIPTION
**Pull request by ZaidMazen1**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/prime_finder.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    # check if a and b are integers
+    assert isinstance(a, int) and isinstance(b, int), "Both a and b must be integers"
+
+    # check if a and b are positive
+    if a < 0 or b < 0:
+        raise ValueError("Integers must be positive")
+
+    # check if a is smaller than b
+    if a > b:
+        raise ValueError("The first integer must be smaller than the second input")
```
Comment (marshar007007) [inline solutions/prime_finder.py]: Your challenge is going well! However, I would to suggest one thing in line number 55 and 56, 
it would be better if you can change that 
if a > b:
    raise ValueError("The start of the range must be less than or equal to the end.") 

Ps: It is just my ideas, without correction to this line, your challenge is also fine!!

File: solutions/prime_finder.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+Module contents:
+  - prime_finder: returns a list of all primes from a list of 2 integers
+
+Created on 2025/01/06
+@author: ZaidMazen1
+"""
+
+
+# defining the function
+def prime_finder(a: int, b: int) -> list[int]:
```
Comment (marshar007007) [inline solutions/prime_finder.py]: In this line, how about changing rename variables **a** and **b** to **start** and **end** to have clearer and simpler variables.

Comment (ZaidMazen1) [reply solutions/prime_finder.py]: Thanks for the recommendation, I've changed up my code to reflect this.

File: solutions/prime_finder.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    # check if a and b are integers
+    assert isinstance(start, int) and isinstance(
+        end, int
+    ), "Both a and b must be integers"
+
+    # check if a and b are positive
+    if start < 0 or end < 0:
+        raise ValueError("Integers must be positive")
+
+    # check if a is smaller than b
```
Comment (marshar007007) [inline solutions/prime_finder.py]: in this line, I think you need to change **a** and **b** to **start** and **end** too

File: solutions/prime_finder.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    ValueError
+
+    >>> prime_finder('cat','dog')
+    AssertionError
+    """
+
+    # check if start and end are integers
+    assert isinstance(start, int) and isinstance(
+        end, int
+    ), "Both a and b must be integers"
```
Comment (marshar007007) [inline solutions/prime_finder.py]: start and end must be integers?

**Review by marshar007007 — APPROVED** [review]
Your challenge is good now! I will approve it.
